File: client-js/src/main/js/dcc/EntitySet.js
/*
=================================================================
personium.io
Copyright 2014 FUJITSU LIMITED

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and limitations under the License.
=================================================================
 */
/*global dcc:false */

///**
//* @class OData関連の各機能を生成/削除するためのクラスの抽象クラス.
//* @constructor
//*/
/**
 * It creates a new object dcc.EntitySet.
 * @class This is the abstract class for performing the merge functions.
 * @param {dcc.Accessor} Accessor
 * @param {dcc.DcCollection} collection
 * @param {String} name
 */
dcc.EntitySet = function(as, col, name) {
  this.initializeProperties(this, as, col, name);
};
dcc.DcClass.inherit(dcc.EntitySet, dcc.box.odata.ODataManager);

///**
//* プロパティを初期化する.
//* @param {dcc.AbstractODataContext} self
//* @param {dcc.Accessor} as アクセス主体
//* @param {?} col
//* @param {?} name
//*/
/**
 * This method initializes the properties of this class.
 * @param {dcc.AbstractODataContext} self
 * @param {dcc.Accessor} as Accessor
 * @param {dcc.DcCollection} col
 * @param {String} name
 */
dcc.EntitySet.prototype.initializeProperties = function(self, as, col, name) {
  this.uber = dcc.box.odata.ODataManager.prototype;
  this.uber.initializeProperties(self, as, col, name);
};

///**
//* Odataデータを部分更新.
//* @param {String} id 部分更新するOdataデータのID値
//* @param {Object} body 部分更新するリクエストボディ
//* @param {String} etag Etag値
//* @throws {ClientException} DAO例外
//*/
/**
 * This method performs the partial update of the Odata data.
 * @param {String} id ID value of the data
 * @param {Object} body Request body
 * @param {String} etag Etag value
 * @param {Object} options optional callback, header and async parameters
 * @throws {dcc.ClientException} DAO exception
 */
dcc.EntitySet.prototype.internalMerge = function(id, body, etag, options) {
  var url = this.getUrl() + "('" + id + "')";
  var rest = dcc.http.RestAdapterFactory.create(this.accessor);
  /*if (options !== undefined) {
    rest.merge(url, JSON.stringify(body), etag, "application/json", function(resp) {
        options(resp);
    });
    rest.merge(url, JSON.stringify(body), etag, "application/json", options);
  } else {*/
    return rest.merge(url, JSON.stringify(body), etag, "application/json", options);
  //}
};

///**
//* Odataデータを部分更新.
//* @param {String} id 対象となるID値
//* @param {Object} body PUTするリクエストボディ
//* @param {String} etag ETag値
//* @throws {ClientException} DAO例外
//*/
/**
 * This method is exposed to perform the partial update of the Odata data.
 * @param {String} id ID value of the data
 * @param {Object} body Request body
 * @param {String} etag ETag value
 * @param {Object} options optional callback, header and async parameters
 * @throws {dcc.ClientException} DAO exception
 */
dcc.EntitySet.prototype.merge = function(id, body, etag, options) {
  //if (options !== undefined) {
    var async = this._isAsynchronous(options);
    var response = this.internalMerge(id, body, etag,options);
    var thenable = null;
    if (async) {
        thenable = response;
    } else {
        thenable = response.resolvedValue;
    }
    return thenable;
    /*this.internalMerge(id, body, etag, function(resp) {
      if (resp.getStatusCode() >= 300) {
        if (options.error !== undefined) {
          options.error(resp);
        }
      } else {
        var odataResponse = new dcc.box.odata.ODataResponse(resp.accessor, "");
        if (options.success !== undefined) {
          options.success(odataResponse);
        }
      }
      if (options.complete !== undefined) {
        options.complete(resp);
      }
    });*/
  /*} else {
    this.internalMerge(id, body, etag);
    return new dcc.box.odata.ODataResponse(this.accessor, "");
  }*/
};

///**
//* Odataデータを取得.
//* @param {String} id 対象となるID値
//* @throws {ClientException} DAO例外
//*/
/**
 * This method gets Odata as response.
 * @param {String} id ID value
 * @param {Object} options optional callback, header and async parameters
 * @throws {dcc.ClientException} DAO exception
 */
dcc.EntitySet.prototype.retrieveAsResponse = function(id, options) {
    var async = this._isAsynchronous(options);
    var response = this._internalRetrieve(id,options);
    var thenable = null;
    if (async) {
        thenable = response;
    } else {
        thenable = response.resolvedValue;
    }
    return thenable;
  /*  if (callback !== undefined) {
    this._internalRetrieve(id, function(resp) {
      if (resp.getStatusCode() >= 300) {
        if (callback.error !== undefined) {
          callback.error(resp);
        }
      } else {
        var responseBody = resp.bodyAsJson();
        var json = responseBody.d.results;
        var odataResponse = new dcc.box.odata.ODataResponse(resp.accessor, json);
        if (callback.success !== undefined) {
          callback.success(odataResponse);
        }
      }
      if (callback.complete !== undefined) {
        callback.complete(resp);
      }
    });
  } */
 // if(this.accessor.getContext().getAsync()){
    //return this._internalRetrieve(id,options);
  //}
  /*else {
    var body = this._internalRetrieve(id);
    return new dcc.box.odata.ODataResponse(this.accessor, body);
  }*/
};

File: client-java/src/main/java/com/fujitsu/dc/client/EntitySet.java
/**
 * personium.io
 * Copyright 2014 FUJITSU LIMITED
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.fujitsu.dc.client;

import java.util.HashMap;

import org.json.simple.JSONObject;

import com.fujitsu.dc.client.http.DcResponse;
import com.fujitsu.dc.client.http.IRestAdapter;
import com.fujitsu.dc.client.http.RestAdapter;
import com.fujitsu.dc.client.http.RestAdapterFactory;
import com.fujitsu.dc.client.utils.Utils;

///**
// * OData関連の各機能を生成/削除するためのクラスの抽象クラス.
// */
/**
 * It creates a new object of EntitySet. This is the abstract class for performing the merge functions.
 */
public class EntitySet extends ODataManager {

    // /**
    // * コンストラクタ.
    // * @param as アクセス主体
    // */
    /**
     * This is the parameterized constructor with one argument and calls its parent constructor internally.
     * @param as Accessor
     */
    public EntitySet(Accessor as) {
        super(as);
    }

    // /**
    // * コンストラクタ.
    // * @param as アクセス主体
    // * @param col DAVコレクション
    // */
    /**
     * This is the parameterized constructor with two arguments and calls its parent constructor internally.
     * @param as Accessor
     * @param col DcCollection
     */
    public EntitySet(Accessor as, DcCollection col) {
        super(as, col);
    }

    // /**
    // * コンストラクタ.
    // * @param as アクセス主体
    // * @param col DAVコレクション
    // * @param name EntitySet名
    // */
    /**
     * This is the parameterized constructor with three arguments and calls its parent constructor internally.
     * @param as Accessor
     * @param col DcCollection
     * @param name EntitySet Name
     */
    public EntitySet(Accessor as, DcCollection col, String name) {
        super(as, col, name);
    }

    // /**
    // * ユーザーデータを部分更新.
    // * @param id 部分更新するOdataデータのID値
    // * @param body 部分更新するリクエストボディ
    // * @param etag Etag値
    // * @throws DaoException DAO例外
    // */
    /**
     * This method is used for the partial update of user data.
     * @param id ID value of the data
     * @param body Request body
     * @param etag Etag value
     * @throws DaoException Exception thrown
     */
    private void internalMerge(String id, HashMap<String, Object> body, String etag) throws DaoException {
        String url = this.getUrl() + "('" + id + "')";
        IRestAdapter rest = RestAdapterFactory.create(accessor);
        rest.merge(url, JSONObject.toJSONString(body), etag, RestAdapter.CONTENT_TYPE_JSON);
    }

    // /**
    // * ユーザーデータを部分更新.
    // * @param id 対象となるID値
    // * @param body PUTするリクエストボディ
    // * @param etag ETag値
    // * @throws DaoException DAO例外
    // */
    /**
     * This method is used for the partial update of user data.
     * @param id ID value of the data
     * @param body Request body to be PUT
     * @param etag ETag value
     * @throws DaoException Exception thrown
     */
    public void merge(String id, HashMap<String, Object> body, String etag) throws DaoException {
        this.internalMerge(id, body, etag);
    }

    // /**
    // * ユーザーデータを部分更新.
    // * @param id 対象となるID値
    // * @param body PUTするリクエストボディ
    // * @param etag ETag値
    // * @return 部分更新したEntityオブジェクト
    // * @throws DaoException DAO例外
    // */
    /**
     * This method is used for the partial update of user data.
     * @param id ID value of the data
     * @param body Request body to be PUT
     * @param etag ETag value
     * @return Partially updated Entity object
     * @throws DaoException Exception thrown
     */
    @SuppressWarnings("unchecked")
    public Entity mergeAsEntity(String id, HashMap<String, Object> body, String etag) throws DaoException {
        String url = this.getUrl() + "('" + id + "')";
        IRestAdapter rest = RestAdapterFactory.create(accessor);
        DcResponse res = rest.merge(url, JSONObject.toJSONString(body), etag, RestAdapter.CONTENT_TYPE_JSON);
        JSONObject json = new JSONObject();
        json.putAll(body);
        Entity entity = new Entity(accessor, json);
        entity.setResHeaders(res.getHeaderList());
        return entity;
    }

    // /**
    // * ユーザーデータを作成.
    // * @param body リクエストボディ
    // * @return 作成したEntityオブジェクト
    // * @throws DaoException DAO例外
    // */
    /**
     * This method is used to create the user data.
     * @param body Request body
     * @return Entity object that is created
     * @throws DaoException Exception thrown
     */
    public Entity createAsEntity(HashMap<String, Object> body) throws DaoException {
        String url = this.getUrl();
        IRestAdapter rest = RestAdapterFactory.create(accessor);
        DcResponse res = rest.post(url, JSONObject.toJSONString(body), RestAdapter.CONTENT_TYPE_JSON);
        JSONObject resbody = (JSONObject) ((JSONObject) res.bodyAsJson().get("d")).get("results");
        Entity entity = new Entity(accessor, (JSONObject) resbody);
        entity.setBody(resbody);
        entity.setResHeaders(res.getHeaderList());
        return entity;
    }

    // /**
    // * ユーザーデータを更新.
    // * @param key 更新するユーザデータのid
    // * @param body リクエストボディ
    // * @param etag 更新対象のバージョンのetag
    // * @return 作成したEntityオブジェクト
    // * @throws DaoException DAO例外
    // */
    /**
     * This method is used to update the user data.
     * @param key ID value of the user data
     * @param body Request body
     * @param etag Etag value
     * @return Entity object that is updated
     * @throws DaoException Exception thrown
     */
    @SuppressWarnings("unchecked")
    public Entity updateAsEntity(String key, HashMap<String, Object> body, String etag) throws DaoException {
        String escapekey = "'" + Utils.escapeURI(key) + "'";
        String url = this.getUrl() + "(" + escapekey + ")";
        IRestAdapter rest = RestAdapterFactory.create(accessor);
        DcResponse res = rest.put(url, JSONObject.toJSONString(body), etag, RestAdapter.CONTENT_TYPE_JSON);
        JSONObject json = new JSONObject();
        json.putAll(body);
        Entity entity = new Entity(accessor, json);
        entity.setResHeaders(res.getHeaderList());
        return entity;
    }

    // /**
    // * ユーザーデータを更新.
    // * @param id 更新するユーザデータのid
    // * @return 作成したEntityオブジェクト
    // * @throws DaoException DAO例外
    // */
    /**
     * This method is used to retrieve the specified user data.
     * @param id ID value of the user data
     * @return Entity object
     * @throws DaoException Exception thrown
     */
    public Entity retrieveAsEntity(String id) throws DaoException {
        DcResponse res = internalRetrieveAsDcResponse(id);
        JSONObject json = (JSONObject) ((JSONObject) res.bodyAsJson().get("d")).get("results");
        Entity entity = new Entity(accessor, json);
        entity.setResHeaders(res.getHeaderList());
        return entity;
    }
}
